flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/BufferDecompressor.java

```diff
@@ -0,0 +1,106 @@
+
+/**
+ * Decompressor for compressed {@link Buffer}.
+ */
+public class BufferDecompressor {
+
+	/** The intermediate heap buffer for the decompressed data. */
+	private final byte[] heapBuffer;
+
+	/** The backing block decompressor for data decompression. */
+	private final BlockDecompressor blockDecompressor;
+
+	public BufferDecompressor(int bufferSize, String factoryName) {
+		checkArgument(bufferSize > 0);
+		this.heapBuffer = new byte[bufferSize];
+		this.blockDecompressor = BlockCompressionFactory.createBlockCompressionFactory(factoryName).getDecompressor();
+	}
+
+	/**
+	 * Decompresses the given {@link Buffer} using {@link BlockDecompressor}. The decompressed data will be stored
+	 * in the internal heap buffer of this {@link BufferDecompressor} and returned to the caller. The caller must
+	 * guarantee that the returned {@link Buffer} is freed when calling the method next time.
+	 *
+	 * <p>Notes that the decompression will always start from offset 0 to the size of the input {@link Buffer}.
+	 */
+	public Buffer decompressToInternalBuffer(Buffer buffer) {
+		int decompressedLen = decompress(buffer);
+
+		// warp the internal heap buffer as Buffer
+		MemorySegment segment = MemorySegmentFactory.wrap(heapBuffer);
+		NetworkBuffer uncompressedBuffer = new NetworkBuffer(segment, FreeingBufferRecycler.INSTANCE);
+		uncompressedBuffer.setSize(decompressedLen);
+		uncompressedBuffer.setCompressed(false);
+
+		return uncompressedBuffer;
+	}
+
+	/**
+	 * The difference between this method and {@link #decompressToInternalBuffer(Buffer)} is that this method copies
+	 * the decompressed data to the input {@link Buffer} starting from offset 0.
+	 *
+	 * <p>The caller must guarantee that the input {@link Buffer} is writable and there's enough space left.
+	 */
+	public Buffer decompressInPlace(Buffer buffer) {
+		int decompressedLen = decompress(buffer);
+
+		// copy the decompressed data back
+		MemorySegment segment = buffer.getMemorySegment();
+		segment.put(buffer.getMemorySegmentOffset(), heapBuffer, 0, decompressedLen);
+		buffer.setCompressed(false);
+		buffer.setSize(decompressedLen);
+
+		return buffer;
+	}
+
+	/**
+	 * Decompresses the input {@link Buffer} into the internal heap buffer and returns the decompressed data size.
+	 */
+	private int decompress(Buffer buffer) {
+		checkArgument(buffer != null, "The input buffer must not be null.");
+		checkArgument(buffer.isCompressed(), "Buffer not compressed.");
+		checkArgument(buffer.getReaderIndex() == 0, "Reader index of the input buffer must be 0.");
+		checkArgument(buffer.readableBytes() > 0, "No data to be decompressed.");
+
+		MemorySegment segment = buffer.getMemorySegment();
+		int length = buffer.getSize();
+		int decompressedLen;
+		// decompress the given buffer into the internal heap buffer
+		if (segment.isOffHeap()) {
+			ByteBuffer src = buffer.getNioBuffer(0, length);
+			decompressedLen = blockDecompressor.decompress(src, 0, length, ByteBuffer.wrap(heapBuffer), 0);
```

Comment:
`return blockDecompressor.decompress(src, 0, length, ByteBuffer.wrap(heapBuffer), 0)` directly, no need to define the local variable.